cpp/src/arrow/engine/expression_test.cc

```diff
@@ -0,0 +1,250 @@
+                                  bool_table.CastTo(i32));
+
+  EXPECT_EQ(bool_scalar.CastTo(i32), ExprType::Scalar(i32));
+  EXPECT_EQ(bool_array.CastTo(i32), ExprType::Array(i32));
+  EXPECT_EQ(bool_table.CastTo(other), ExprType::Table(other));
+}
+
+class ExprTest : public testing::Test {};
+
+TEST_F(ExprTest, ScalarExpr) {
+  ASSERT_RAISES(Invalid, ScalarExpr::Make(nullptr));
+
+  auto i32 = int32();
+  ASSERT_OK_AND_ASSIGN(auto value, MakeScalar(i32, 10));
+  ASSERT_OK_AND_ASSIGN(auto expr, ScalarExpr::Make(value));
+  EXPECT_EQ(expr->kind(), Expr::SCALAR_LITERAL);
+  EXPECT_EQ(expr->type(), ExprType::Scalar(i32));
+  EXPECT_EQ(*expr->scalar(), *value);
+}
+
+TEST_F(ExprTest, FieldRefExpr) {
+  ASSERT_RAISES(Invalid, FieldRefExpr::Make(nullptr));
+
+  auto i32 = int32();
+  auto f_i32 = field("i32", i32);
+
+  ASSERT_OK_AND_ASSIGN(auto expr, FieldRefExpr::Make(f_i32));
+  EXPECT_EQ(expr->kind(), Expr::FIELD_REFERENCE);
+  EXPECT_EQ(expr->type(), ExprType::Array(i32));
+  EXPECT_THAT(expr->field(), PtrEquals(f_i32));
+}
+
+template <typename CmpClass>
+class CmpExprTest : public ExprTest {
+ public:
+  Expr::Kind kind() { return expr_traits<CmpClass>::kind_id; }
+
+  Result<std::shared_ptr<CmpClass>> Make(std::shared_ptr<Expr> left,
+                                         std::shared_ptr<Expr> right) {
+    return CmpClass::Make(std::move(left), std::move(right));
+  }
+};
+
+using CompareExprs = ::testing::Types<EqualCmpExpr, NotEqualCmpExpr>;
+
+TYPED_TEST_CASE(CmpExprTest, CompareExprs);
+TYPED_TEST(CmpExprTest, BasicCompareExpr) {
+  auto i32 = int32();
+
+  auto f_i32 = field("i32", i32);
+  ASSERT_OK_AND_ASSIGN(auto f_expr, FieldRefExpr::Make(f_i32));
+
+  ASSERT_OK_AND_ASSIGN(auto s_i32, MakeScalar(i32, 42));
+  ASSERT_OK_AND_ASSIGN(auto s_expr, ScalarExpr::Make(s_i32));
+
+  // Required fields
+  ASSERT_RAISES(Invalid, this->Make(nullptr, nullptr));
+  ASSERT_RAISES(Invalid, this->Make(s_expr, nullptr));
+  ASSERT_RAISES(Invalid, this->Make(nullptr, f_expr));
+
+  // Not type compatible
+  ASSERT_OK_AND_ASSIGN(auto s_i64, MakeScalar(int64(), 42L));
+  ASSERT_OK_AND_ASSIGN(auto s_expr_i64, ScalarExpr::Make(s_i64));
+  EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid, HasSubstr("operands must be of same type"),
+                                  this->Make(s_expr_i64, f_expr));
+
+  ASSERT_OK_AND_ASSIGN(auto expr, this->Make(f_expr, s_expr));
+  EXPECT_EQ(expr->kind(), this->kind());
+  // Ensure type is broadcasted
+  EXPECT_EQ(expr->type(), ExprType::Array(boolean()));
+  EXPECT_TRUE(expr->type().IsPredicate());
+  EXPECT_THAT(expr, PtrEquals(expr));
+  EXPECT_THAT(expr->left_operand(), PtrEquals(f_expr));
+  EXPECT_THAT(expr->right_operand(), PtrEquals(s_expr));
+
+  ASSERT_OK_AND_ASSIGN(auto other, this->Make(f_expr, s_expr));
+  EXPECT_THAT(expr, PtrEquals(other));
+  // Compare operators supports commutativity
+  // TODO(fsaintjacques): what about floating point types?
```

Comment:
Well, why do we need to handle commutativity in `Expr::Equals`? Is there any point in that?